File: src/lib/ai/neural-embeddings.ts
import { invoke } from '@tauri-apps/api/core';

export interface NeuralEmbeddingResult {
  embedding: number[];
  text: string;
  memoryType?: string;
  similarity?: number;
}

export interface NeuralEmbeddingSearchResult {
  text: string;
  similarity: number;
  memoryType?: string;
  metadata?: Record<string, unknown>;
}

export interface NeuralEmbeddingConfig {
  embeddingDim?: number;
  maxTextLength?: number;
  learningRate?: number;
  trainingEpochs?: number;
  cacheSizeLimit?: number;
}

export interface NeuralEmbeddingStats {
  cacheSize: number;
  cacheLimit: number;
  embeddingDimension: number;
  specializedNetworks: number;
}

/**
 * Neural embedding service that uses the Rust backend's neural networks
 * for generating meaningful embeddings for different memory types
 */
export class NeuralEmbeddingService {
  private initialized = false;

  constructor(private config: NeuralEmbeddingConfig = {}) {}

  /**
   * Initialize the neural embedding service
   */
  async initialize(): Promise<void> {
    if (this.initialized) return;

    try {
      await invoke('init_neural_embedding_service', {
        config: this.config,
      });
      this.initialized = true;
    } catch (error) {
      throw new Error(`Failed to initialize neural embedding service: ${error}`);
    }
  }

  /**
   * Generate embedding for a single text
   */
  async generateEmbedding(
    text: string,
    memoryType?: string
  ): Promise<NeuralEmbeddingResult> {
    await this.ensureInitialized();

    try {
      const result = await invoke<{
        embedding: number[];
        text: string;
        memoryType?: string;
      }>('generate_neural_embedding', {
        text,
        memoryType,
      });

      return {
        embedding: result.embedding,
        text: result.text,
        memoryType: result.memoryType,
      };
    } catch (error) {
      throw new Error(`Failed to generate neural embedding: ${error}`);
    }
  }

  /**
   * Generate embeddings for multiple texts
   */
  async generateEmbeddings(
    texts: Array<{ text: string; memoryType?: string }>
  ): Promise<NeuralEmbeddingResult[]> {
    await this.ensureInitialized();

    try {
      const results = await invoke<Array<{
        embedding: number[];
        text: string;
        memoryType?: string;
      }>>('generate_neural_embeddings_batch', {
        texts,
      });

      return results.map((result) => ({
        embedding: result.embedding,
        text: result.text,
        memoryType: result.memoryType,
      }));
    } catch (error) {
      throw new Error(`Failed to generate neural embeddings: ${error}`);
    }
  }

  /**
   * Calculate cosine similarity between two embeddings
   */
  calculateSimilarity(embedding1: number[], embedding2: number[]): number {
    if (embedding1.length !== embedding2.length) {
      return 0.0;
    }

    const dotProduct = embedding1.reduce((sum, val, i) => sum + val * embedding2[i], 0);
    const norm1 = Math.sqrt(embedding1.reduce((sum, val) => sum + val * val, 0));
    const norm2 = Math.sqrt(embedding2.reduce((sum, val) => sum + val * val, 0));

    if (norm1 === 0 || norm2 === 0) {
      return 0;
    }

    return dotProduct / (norm1 * norm2);
  }

  /**
   * Search for similar embeddings using neural networks
   */
  async searchSimilar(
    queryEmbedding: number[],
    candidateEmbeddings: Array<{
      embedding: number[];
      text: string;
      memoryType?: string;
      metadata?: Record<string, unknown>;
    }>,
    topK = 5,
    threshold = 0.7
  ): Promise<NeuralEmbeddingSearchResult[]> {
    await this.ensureInitialized();

    try {
      const results = await invoke<Array<{
        text: string;
        similarity: number;
        memoryType?: string;
        metadata?: Record<string, unknown>;
      }>>('search_neural_similar', {
        queryEmbedding,
        candidateEmbeddings,
        topK,
        threshold,
      });

      return results;
    } catch (error) {
      throw new Error(`Failed to search similar embeddings: ${error}`);
    }
  }

  /**
   * Find similar memories using neural networks
   */
  async findSimilarMemories(
    queryText: string,
    agentId: string,
    memoryType?: string,
    threshold = 0.7,
    topK = 5
  ): Promise<NeuralEmbeddingSearchResult[]> {
    await this.ensureInitialized();

    try {
      const results = await invoke<Array<{
        text: string;
        similarity: number;
        memoryType?: string;
        metadata?: Record<string, unknown>;
      }>>('find_similar_memories', {
        queryText,
        agentId,
        memoryType,
        threshold,
        topK,
      });

      return results;
    } catch (error) {
      throw new Error(`Failed to find similar memories: ${error}`);
    }
  }

  /**
   * Train the neural networks on new data
   */
  async trainOnMemories(agentId: string): Promise<void> {
    await this.ensureInitialized();

    try {
      await invoke('train_neural_networks', {
        agentId,
      });
    } catch (error) {
      throw new Error(`Failed to train neural networks: ${error}`);
    }
  }

  /**
   * Get embedding service statistics
   */
  async getStats(): Promise<NeuralEmbeddingStats> {
    await this.ensureInitialized();

    try {
      const stats = await invoke<NeuralEmbeddingStats>('get_neural_embedding_stats');
      return stats;
    } catch (error) {
      throw new Error(`Failed to get neural embedding stats: ${error}`);
    }
  }

  /**
   * Clear the embedding cache
   */
  async clearCache(): Promise<void> {
    await this.ensureInitialized();

    try {
      await invoke('clear_neural_embedding_cache');
    } catch (error) {
      throw new Error(`Failed to clear neural embedding cache: ${error}`);
    }
  }

  /**
   * Ensure the service is initialized
   */
  private async ensureInitialized(): Promise<void> {
    if (!this.initialized) {
      await this.initialize();
    }
  }
}

/**
 * Default neural embedding service instance
 */
export const neuralEmbeddingService = new NeuralEmbeddingService();

/**
 * Utility functions for common neural embedding tasks
 */

/**
 * Generate neural embedding for text search
 */
export async function generateNeuralSearchEmbedding(
  query: string,
  memoryType?: string
): Promise<number[]> {
  const result = await neuralEmbeddingService.generateEmbedding(query, memoryType);
  return result.embedding;
}

/**
 * Find similar texts using neural embeddings
 */
export async function findSimilarTextsNeural(
  query: string,
  texts: Array<{ text: string; memoryType?: string }>,
  topK = 5,
  threshold = 0.7
): Promise<NeuralEmbeddingSearchResult[]> {
  // Generate embeddings for all texts
  const [queryResult, textResults] = await Promise.all([
    neuralEmbeddingService.generateEmbedding(query),
    neuralEmbeddingService.generateEmbeddings(texts),
  ]);

  // Prepare candidates
  const candidates = textResults.map((result) => ({
    embedding: result.embedding,
    text: result.text,
    memoryType: result.memoryType,
  }));

  // Search for similar texts
  return neuralEmbeddingService.searchSimilar(
    queryResult.embedding,
    candidates,
    topK,
    threshold
  );
}

/**
 * Cluster texts by neural similarity
 */
export async function clusterTextsNeural(
  texts: Array<{ text: string; memoryType?: string }>,
  similarityThreshold = 0.8
): Promise<Array<{ text: string; memoryType?: string }>[]> {
  if (texts.length === 0) return [];

  const embeddings = await neuralEmbeddingService.generateEmbeddings(texts);
  const clusters: Array<{ text: string; memoryType?: string }>[][] = [];
  const processed = new Set<number>();

  for (let i = 0; i < embeddings.length; i++) {
    if (processed.has(i)) continue;

    const currentEmbedding = embeddings[i];
    if (!currentEmbedding) continue;

    const cluster = [{
      text: currentEmbedding.text,
      memoryType: currentEmbedding.memoryType,
    }];
    processed.add(i);

    for (let j = i + 1; j < embeddings.length; j++) {
      if (processed.has(j)) continue;

      const otherEmbedding = embeddings[j];
      if (!otherEmbedding) continue;

      const similarity = neuralEmbeddingService.calculateSimilarity(
        currentEmbedding.embedding,
        otherEmbedding.embedding
      );

      if (similarity >= similarityThreshold) {
        cluster.push({
          text: otherEmbedding.text,
          memoryType: otherEmbedding.memoryType,
        });
        processed.add(j);
      }
    }

    clusters.push(cluster);
  }

  return clusters;
}

/**
 * Create a semantic index using neural embeddings
 */
export class NeuralSemanticIndex {
  private embeddings: Array<{
    id: string;
    text: string;
    embedding: number[];
    memoryType?: string;
    metadata?: Record<string, unknown>;
  }> = [];

  constructor(private embeddingService: NeuralEmbeddingService = new NeuralEmbeddingService()) {}

  /**
   * Add text to the index
   */
  async addText(
    id: string,
    text: string,
    memoryType?: string,
    metadata?: Record<string, unknown>
  ): Promise<void> {
    const result = await this.embeddingService.generateEmbedding(text, memoryType);

    this.embeddings.push({
      id,
      text,
      embedding: result.embedding,
      memoryType: result.memoryType,
      ...(metadata && { metadata }),
    });
  }

  /**
   * Add multiple texts to the index
   */
  async addTexts(
    items: Array<{
      id: string;
      text: string;
      memoryType?: string;
      metadata?: Record<string, unknown>;
    }>
  ): Promise<void> {
    const texts = items.map((item) => ({
      text: item.text,
      memoryType: item.memoryType,
    }));
    const results = await this.embeddingService.generateEmbeddings(texts);

    items.forEach((item, index) => {
      const result = results[index];
      if (result) {
        this.embeddings.push({
          id: item.id,
          text: item.text,
          embedding: result.embedding,
          memoryType: result.memoryType,
          ...(item.metadata && { metadata: item.metadata }),
        });
      }
    });
  }

  /**
   * Search the index
   */
  async search(
    query: string,
    topK = 5,
    threshold = 0.7
  ): Promise<
    Array<{
      id: string;
      text: string;
      similarity: number;
      memoryType?: string;
      metadata?: Record<string, unknown>;
    }>
  > {
    const queryResult = await this.embeddingService.generateEmbedding(query);

    const results = this.embeddings
      .map((item) => ({
        id: item.id,
        text: item.text,
        similarity: this.embeddingService.calculateSimilarity(
          queryResult.embedding,
          item.embedding
        ),
        memoryType: item.memoryType,
        ...(item.metadata && { metadata: item.metadata }),
      }))
      .filter((result) => result.similarity >= threshold)
      .sort((a, b) => b.similarity - a.similarity)
      .slice(0, topK);

    return results;
  }

  /**
   * Remove text from index
   */
  removeText(id: string): boolean {
    const index = this.embeddings.findIndex((item) => item.id === id);
    if (index !== -1) {
      this.embeddings.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Clear the index
   */
  clear(): void {
    this.embeddings = [];
  }

  /**
   * Get index size
   */
  size(): number {
    return this.embeddings.length;
  }

  /**
   * Export index data
   */
  export(): Array<{
    id: string;
    text: string;
    embedding: number[];
    memoryType?: string;
    metadata?: Record<string, unknown>;
  }> {
    return [...this.embeddings];
  }

  /**
   * Import index data
   */
  import(
    data: Array<{
      id: string;
      text: string;
      embedding: number[];
      memoryType?: string;
      metadata?: Record<string, unknown>;
    }>
  ): void {
    this.embeddings = [...data];
  }
} 

File: src-tauri/src/database/neural_embeddings.rs
use anyhow::{Result, anyhow};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use sha2::{Sha256, Digest};
use super::neural_network::{NeuralNetwork, NetworkBuilder, ActivationFunction, TrainingData};
use super::memory::{MemoryType, AgentMemory};
use serde::{Serialize, Deserialize};

/// Neural embedding service that uses FANN-inspired neural networks
/// to generate meaningful embeddings for different memory types
pub struct NeuralEmbeddingService {
    /// Cache for computed embeddings
    cache: Arc<RwLock<HashMap<String, Vec<f32>>>>,
    /// Neural networks specialized for different memory types
    memory_networks: HashMap<MemoryType, NeuralNetwork>,
    /// General purpose embedding network
    general_network: NeuralNetwork,
    /// Configuration for the service
    config: EmbeddingConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmbeddingConfig {
    pub embedding_dim: usize,
    pub max_text_length: usize,
    pub learning_rate: f32,
    pub training_epochs: usize,
    pub cache_size_limit: usize,
}

impl Default for EmbeddingConfig {
    fn default() -> Self {
        Self {
            embedding_dim: 256,
            max_text_length: 512,
            learning_rate: 0.001,
            training_epochs: 100,
            cache_size_limit: 10000,
        }
    }
}

impl NeuralEmbeddingService {
    /// Create a new neural embedding service
    pub async fn new(config: Option<EmbeddingConfig>) -> Result<Self> {
        let config = config.unwrap_or_default();
        
        // Create general purpose embedding network
        let general_network = NetworkBuilder::new()
            .input_layer(config.max_text_length) // Text feature input
            .hidden_layer_with_activation(512, ActivationFunction::ReLU, 0.1)
            .hidden_layer_with_activation(384, ActivationFunction::GELU, 0.1)
            .output_layer(config.embedding_dim)
            .learning_rate(config.learning_rate)
            .build()?;

        // Create specialized networks for different memory types
        let mut memory_networks = HashMap::new();
        
        // Conversation memory network - optimized for dialogue understanding
        let conversation_network = NetworkBuilder::new()
            .input_layer(config.max_text_length)
            .hidden_layer_with_activation(256, ActivationFunction::Tanh, 0.1)
            .hidden_layer_with_activation(128, ActivationFunction::ReLU, 0.05)
            .output_layer(config.embedding_dim)
            .learning_rate(config.learning_rate * 1.2) // Slightly higher learning rate
            .build()?;
        memory_networks.insert(MemoryType::Conversation, conversation_network);

        // Task memory network - optimized for action/goal understanding
        let task_network = NetworkBuilder::new()
            .input_layer(config.max_text_length)
            .hidden_layer_with_activation(384, ActivationFunction::ReLU, 0.1)
            .hidden_layer_with_activation(192, ActivationFunction::GELU, 0.1)
            .output_layer(config.embedding_dim)
            .learning_rate(config.learning_rate)
            .build()?;
        memory_networks.insert(MemoryType::Task, task_network);

        // Learning memory network - optimized for knowledge extraction
        let learning_network = NetworkBuilder::new()
            .input_layer(config.max_text_length)
            .hidden_layer_with_activation(512, ActivationFunction::GELU, 0.15)
            .hidden_layer_with_activation(256, ActivationFunction::Tanh, 0.1)
            .output_layer(config.embedding_dim)
            .learning_rate(config.learning_rate * 0.8) // Slower learning for stability
            .build()?;
        memory_networks.insert(MemoryType::Learning, learning_network);

        // Pattern memory network - optimized for pattern recognition
        let pattern_network = NetworkBuilder::new()
            .input_layer(config.max_text_length)
            .hidden_layer_with_activation(128, ActivationFunction::ReLU, 0.2)
            .hidden_layer_with_activation(64, ActivationFunction::Sigmoid, 0.1)
            .output_layer(config.embedding_dim)
            .learning_rate(config.learning_rate * 1.5)
            .build()?;
        memory_networks.insert(MemoryType::Pattern, pattern_network);

        Ok(Self {
            cache: Arc::new(RwLock::new(HashMap::new())),
            memory_networks,
            general_network,
            config,
        })
    }

    /// Generate embedding for text using appropriate neural network
    pub async fn embed_text(&self, text: &str, memory_type: Option<MemoryType>) -> Result<Vec<f32>> {
        // Check cache first
        let cache_key = self.generate_cache_key(text, &memory_type);
        
        {
            let cache = self.cache.read().await;
            if let Some(cached_embedding) = cache.get(&cache_key) {
                return Ok(cached_embedding.clone());
            }
        }

        // Preprocess text into neural network input
        let input_features = self.text_to_features(text)?;
        
        // Select appropriate network
        let embedding = if let Some(ref mem_type) = memory_type {
            if let Some(network) = self.memory_networks.get(mem_type) {
                network.run(&input_features)
            } else {
                self.general_network.run(&input_features)
            }
        } else {
            self.general_network.run(&input_features)
        };

        // Normalize the embedding
        let normalized_embedding = self.normalize_embedding(&embedding);

        // Cache the result (with size limit)
        {
            let mut cache = self.cache.write().await;
            if cache.len() >= self.config.cache_size_limit {
                // Remove oldest entries (simple LRU would be better, but this works)
                cache.clear();
            }
            cache.insert(cache_key, normalized_embedding.clone());
        }

        Ok(normalized_embedding)
    }

    /// Generate embeddings for multiple texts
    pub async fn embed_batch(
        &self, 
        texts: &[(String, Option<MemoryType>)]
    ) -> Result<Vec<Vec<f32>>> {
        let mut embeddings = Vec::new();
        
        for (text, memory_type) in texts {
            let embedding = self.embed_text(text, memory_type.clone()).await?;
            embeddings.push(embedding);
        }

        Ok(embeddings)
    }

    /// Generate embedding specifically for agent memory
    pub async fn embed_memory(&self, memory: &AgentMemory) -> Result<Vec<f32>> {
        // Combine content with metadata for richer embeddings
        let enhanced_text = self.enhance_text_with_metadata(memory);
        self.embed_text(&enhanced_text, Some(memory.memory_type.clone())).await
    }

    /// Train networks on memory data to improve embeddings
    pub async fn train_on_memories(&mut self, memories: &[AgentMemory]) -> Result<()> {
        if memories.is_empty() {
            return Ok(());
        }

        // Group memories by type for specialized training
        let mut memory_groups: HashMap<MemoryType, Vec<&AgentMemory>> = HashMap::new();
        for memory in memories {
            memory_groups.entry(memory.memory_type.clone())
                .or_insert_with(Vec::new)
                .push(memory);
        }

        // Prepare all training data first to avoid borrowing conflicts
        let mut prepared_training_data = HashMap::new();
        
        // Prepare training data for each memory type
        for (memory_type, group_memories) in &memory_groups {
            let training_data = self.prepare_training_data(group_memories)?;
            prepared_training_data.insert(memory_type.clone(), training_data);
        }
        
        // Prepare training data for general network
        let memory_refs: Vec<&AgentMemory> = memories.iter().collect();
        let all_training_data = self.prepare_training_data(&memory_refs)?;

        // Train each specialized network
        for (memory_type, training_data) in prepared_training_data {
            if let Some(network) = self.memory_networks.get_mut(&memory_type) {
                Self::train_network_static(network, &training_data, self.config.training_epochs, self.config.learning_rate).await?;
            }
        }

        // Train general network on all memories
        Self::train_network_static(&mut self.general_network, &all_training_data, self.config.training_epochs, self.config.learning_rate).await?;

        // Clear cache after training since embeddings will have changed
        {
            let mut cache = self.cache.write().await;
            cache.clear();
        }

        Ok(())
    }

    /// Compute similarity between two embeddings
    pub fn compute_similarity(&self, embedding1: &[f32], embedding2: &[f32]) -> f32 {
        cosine_similarity(embedding1, embedding2)
    }

    /// Find similar embeddings from a set of candidates
    pub async fn find_similar_memories(
        &self,
        query_text: &str,
        query_type: Option<MemoryType>,
        candidate_memories: &[AgentMemory],
        threshold: f32,
        top_k: usize,
    ) -> Result<Vec<(String, f32)>> {
        let query_embedding = self.embed_text(query_text, query_type).await?;
        
        let mut similarities = Vec::new();
        
        for memory in candidate_memories {
            let memory_embedding = self.embed_memory(memory).await?;
            let similarity = self.compute_similarity(&query_embedding, &memory_embedding);
            
            if similarity >= threshold {
                similarities.push((memory.id.clone(), similarity));
            }
        }

        // Sort by similarity (descending)
        similarities.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        // Take top k results
        similarities.truncate(top_k);
        Ok(similarities)
    }

    /// Get embedding statistics
    pub async fn get_stats(&self) -> EmbeddingStats {
        let cache = self.cache.read().await;
        EmbeddingStats {
            cache_size: cache.len(),
            cache_limit: self.config.cache_size_limit,
            embedding_dimension: self.config.embedding_dim,
            specialized_networks: self.memory_networks.len(),
        }
    }

    // Private helper methods

    /// Convert text to neural network input features
    fn text_to_features(&self, text: &str) -> Result<Vec<f32>> {
        let mut features = vec![0.0; self.config.max_text_length];
        
        // Character-level encoding with position awareness
        let chars: Vec<char> = text.chars().collect();
        for (i, &ch) in chars.iter().enumerate() {
            if i >= self.config.max_text_length {
                break;
            }
            
            // Normalize character to 0-1 range
            let char_value = (ch as u32 as f32) / 65536.0; // Unicode max
            features[i] = char_value;
        }

        // Add text statistics as features in remaining positions
        let text_len = chars.len();
        if text_len + 10 < self.config.max_text_length {
            let stats_start = self.config.max_text_length - 10;
            
            // Text length (normalized)
            features[stats_start] = (text_len as f32) / 1000.0;
            
            // Word count
            let word_count = text.split_whitespace().count();
            features[stats_start + 1] = (word_count as f32) / 100.0;
            
            // Average word length
            let avg_word_len = if word_count > 0 {
                text_len as f32 / word_count as f32
            } else {
                0.0
            };
            features[stats_start + 2] = avg_word_len / 20.0;
            
            // Punctuation density
            let punct_count = text.chars().filter(|c| c.is_ascii_punctuation()).count();
            features[stats_start + 3] = (punct_count as f32) / (text_len as f32 + 1.0);
            
            // Uppercase ratio
            let upper_count = text.chars().filter(|c| c.is_uppercase()).count();
            features[stats_start + 4] = (upper_count as f32) / (text_len as f32 + 1.0);
        }

        Ok(features)
    }

    /// Enhance memory text with metadata for richer embeddings
    fn enhance_text_with_metadata(&self, memory: &AgentMemory) -> String {
        let mut enhanced = memory.content.clone();
        
        // Add memory type context
        enhanced.push_str(&format!(" [TYPE:{}]", memory.memory_type));
        
        // Add relevant metadata
        for (key, value) in &memory.metadata {
            enhanced.push_str(&format!(" [{}:{}]", key, value));
        }
        
        // Add tags
        if !memory.tags.is_empty() {
            enhanced.push_str(&format!(" [TAGS:{}]", memory.tags.join(",")));
        }
        
        enhanced
    }

    /// Generate cache key for embeddings
    fn generate_cache_key(&self, text: &str, memory_type: &Option<MemoryType>) -> String {
        let mut hasher = Sha256::new();
        hasher.update(text.as_bytes());
        if let Some(ref mem_type) = memory_type {
            hasher.update(format!("{:?}", mem_type).as_bytes());
        }
        format!("{:x}", hasher.finalize())
    }

    /// Normalize embedding vector
    fn normalize_embedding(&self, embedding: &[f32]) -> Vec<f32> {
        let norm: f32 = embedding.iter().map(|x| x * x).sum::<f32>().sqrt();
        if norm > 0.0 {
            embedding.iter().map(|x| x / norm).collect()
        } else {
            embedding.to_vec()
        }
    }

    /// Prepare training data from memories
    fn prepare_training_data(&self, memories: &[&AgentMemory]) -> Result<TrainingData> {
        let mut training_data = TrainingData::new();
        
        for memory in memories {
            let input_features = self.text_to_features(&self.enhance_text_with_metadata(memory))?;
            
            // Create target embedding based on memory similarity patterns
            // This is a simplified approach - in practice, you'd use more sophisticated targets
            let target_embedding = self.create_target_embedding(memory)?;
            
            training_data.add_example(input_features, target_embedding);
        }
        
        Ok(training_data)
    }

    /// Create target embedding for training
    fn create_target_embedding(&self, memory: &AgentMemory) -> Result<Vec<f32>> {
        let mut target = vec![0.0; self.config.embedding_dim];
        
        // Create semantic targets based on memory characteristics
        let content_hash = self.hash_text(&memory.content);
        let hash_bytes = hex::decode(&content_hash).map_err(|e| anyhow!("Hash decode error: {}", e))?;
        
        // Use hash to create deterministic but distributed target
        for (i, &byte) in hash_bytes.iter().enumerate() {
            if i < self.config.embedding_dim {
                target[i] = (byte as f32) / 255.0;
            }
        }
        
        // Add memory type bias
        let type_bias = match memory.memory_type {
            MemoryType::Conversation => 0.1,
            MemoryType::Task => 0.2,
            MemoryType::Learning => 0.3,
            MemoryType::Pattern => 0.4,
            MemoryType::Context => 0.5,
            MemoryType::Tool => 0.6,
            MemoryType::Error => 0.7,
            MemoryType::Success => 0.8,
        };
        
        if self.config.embedding_dim > 10 {
            target[self.config.embedding_dim - 1] = type_bias;
        }
        
        Ok(self.normalize_embedding(&target))
    }

    /// Train a neural network with training data (static version to avoid borrowing issues)
    async fn train_network_static(
        network: &mut NeuralNetwork, 
        training_data: &TrainingData, 
        epochs: usize,
        _learning_rate: f32
    ) -> Result<()> {
        if training_data.is_empty() {
            return Ok(());
        }

        let errors = network.train(
            &training_data.inputs,
            &training_data.outputs,
            epochs,
        )?;

        // Log training progress (in real implementation, use proper logging)
        let final_error = errors.last().unwrap_or(&f32::INFINITY);
        println!("Neural embedding training completed. Final error: {:.6}", final_error);

        Ok(())
    }

    /// Hash text for deterministic targets
    fn hash_text(&self, text: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(text.as_bytes());
        format!("{:x}", hasher.finalize())
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct EmbeddingStats {
    pub cache_size: usize,
    pub cache_limit: usize,
    pub embedding_dimension: usize,
    pub specialized_networks: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NeuralEmbeddingResult {
    pub embedding: Vec<f32>,
    pub text: String,
    pub memory_type: Option<String>,
    pub similarity: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NeuralEmbeddingSearchResult {
    pub text: String,
    pub similarity: f32,
    pub memory_type: Option<String>,
    pub metadata: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NeuralEmbeddingRequest {
    pub text: String,
    pub memory_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NeuralEmbeddingCandidate {
    pub text: String,
    pub embedding: Vec<f32>,
    pub memory_type: Option<String>,
    pub metadata: Option<HashMap<String, String>>,
}

// Utility functions
pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    if a.len() != b.len() {
        return 0.0;
    }

    let dot_product: f32 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    let norm_a: f32 = a.iter().map(|x| x * x).sum::<f32>().sqrt();
    let norm_b: f32 = b.iter().map(|x| x * x).sum::<f32>().sqrt();

    if norm_a == 0.0 || norm_b == 0.0 {
        0.0
    } else {
        dot_product / (norm_a * norm_b)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;

    #[tokio::test]
    async fn test_neural_embedding_service() {
        let service = NeuralEmbeddingService::new(None).await.unwrap();
        
        let text = "This is a test memory about learning patterns";
        let embedding = service.embed_text(text, Some(MemoryType::Learning)).await.unwrap();
        
        assert_eq!(embedding.len(), 256); // Default embedding dimension
        
        // Test that the embedding is normalized
        let norm: f32 = embedding.iter().map(|x| x * x).sum::<f32>().sqrt();
        assert!((norm - 1.0).abs() < 0.01);
    }

    #[tokio::test]
    async fn test_memory_type_specialization() {
        let service = NeuralEmbeddingService::new(None).await.unwrap();
        
        let text = "Complete the task successfully";
        let task_embedding = service.embed_text(text, Some(MemoryType::Task)).await.unwrap();
        let general_embedding = service.embed_text(text, None).await.unwrap();
        
        // Embeddings should be different due to network specialization
        let similarity = cosine_similarity(&task_embedding, &general_embedding);
        assert!(similarity < 1.0); // Should not be identical
    }

    #[tokio::test]
    async fn test_memory_embedding() {
        let service = NeuralEmbeddingService::new(None).await.unwrap();
        
        let memory = AgentMemory::new(
            "test_agent".to_string(),
            MemoryType::Learning,
            "This is a learning memory".to_string(),
        );
        
        let embedding = service.embed_memory(&memory).await.unwrap();
        assert_eq!(embedding.len(), 256);
    }

    #[test]
    fn test_text_to_features() {
        let config = EmbeddingConfig::default();
        let service = NeuralEmbeddingService {
            cache: Arc::new(RwLock::new(HashMap::new())),
            memory_networks: HashMap::new(),
            general_network: NetworkBuilder::new()
                .input_layer(10)
                .output_layer(5)
                .build()
                .unwrap(),
            config,
        };
        
        let features = service.text_to_features("Hello").unwrap();
        assert_eq!(features.len(), 512); // max_text_length
        assert!(features[0] > 0.0); // Should have some content from 'H'
    }
}